returnn/tf/layers/base.py

```diff
@@ -1271,17 +1271,23 @@ def batch_norm(self, data,
       else:
         sample_variance = updated_sample_variance
       # If train or if force_sample, use default use_sample=0.0, otherwise use_sample=1.0.
-      if self.network.train_flag is not False or force_sample:
-        if force_sample:
-          pass  # leave use_sample as-is
-        else:
-          use_sample = tf.where(self.network.train_flag, use_sample, 1.0)
+      skip_updates = self.network.train_flag is False and update_sample_only_in_training and not force_sample
+      if skip_updates:
+        mean = sample_mean
+        variance = sample_variance
       else:
-        use_sample = 1.0
-      mean = (1. - use_sample) * mean + use_sample * sample_mean
-      variance = (1. - use_sample) * variance + use_sample * sample_variance
-      bn = (data.placeholder - mean) * tf_compat.v1.rsqrt(variance + epsilon)
-      if delayed_ops:
+        if self.network.train_flag is not False or force_sample:
+          if force_sample:
+            pass  # leave use_sample as-is
+          else:
+            use_sample = tf.where(self.network.train_flag, use_sample, 1.0)
+        else:
+          use_sample = 1.0
+        mean = (1. - use_sample) * mean + use_sample * sample_mean
+        variance = (1. - use_sample) * variance + use_sample * sample_variance
+      variance_with_epsilon = variance if epsilon == 0.0 else variance + epsilon
+      bn = (data.placeholder - mean) * tf_compat.v1.rsqrt(variance_with_epsilon)
+      if not skip_updates and delayed_ops and use_sample != 0.0:
```

Comment:
`use_sample != 0.0` is incorrect, as it might be a TF tensor here.

Reply:
I think this is correct. When use_sample is a tensor, we can not optimize this, as it can have any value. So only when we "know" that use_sample is actually fixed 0.0 this step can be skipped. But for the originally intended case this is not needed, so if you want I can remove it.

Reply:
You cannot do `if tf_tensor:` when `tf_tensor` is a `tf.Tensor` because `if` needs a `bool` (or sth which can be casted to `bool`). This is not eager-mode, so it would need a `session.run`.

Reply:
But why would I want to do that in the first place? I do not understand your comment.

Reply:
I did not write this code. I don't want to do that. But the code does this.

`if ... and use_sample != 0.0:` is exactly the case I described.

Reply:
Nowhere in the code is `if tf_tensor`. What I can have is `tf_tensor != 0.0` which is True, and thus it behaves like I want it to behave.

Reply:
> Nowhere in the code is `if tf_tensor`. What I can have is `tf_tensor != 0.0` which is True, and thus it behaves like I want it to behave.

You have this code here:
`if ... and use_sample != 0.0:`

Which is exactly this case. `use_sample != 0.0` is a `tf.Tensor`.

Actually it depends on the TF version, and might raise an exception if `use_sample` is a `tf.Tensor`.

Why are we arguing about this? Just do not do an `if`-check on a `tf.Tensor`.

Write the code in a way that does not do this. E.g. add some `not isinstance(use_sample, tf.Tensor)`. Then it would be fine. (But not sure if this is what you really want here.)

Reply:
Why should I write it more complicated than needed?

TF1:
```
In [10]: import tensorflow as tf

In [11]: tf.compat.v1.disable_eager_execution()

In [12]: a = tf.compat.v1.zeros(())

In [13]: a
Out[13]: <tf.Tensor 'zeros_1:0' shape=() dtype=float32>

In [14]: a != 0.0
Out[14]: True
```

TF2:
```
In [2]: import tensorflow as tf

In [3]: a = tf.zeros(())

In [4]: a
Out[4]: <tf.Tensor: shape=(), dtype=float32, numpy=0.0>

In [5]: a != 0.0
Out[5]: <tf.Tensor: shape=(), dtype=bool, numpy=False>

In [6]: bool (a != 0.0)
Out[6]: False
```

So the check is "can the variable be other than zero". Which is always True for graph mode, because I can not look into the Tensor. And might be "True" in eager mode when I can look inside. So no matter the TF version and mode this will evaluate as I want it to be.

But I do not care about this case that is covered by this (which is `force_sample=True` and `use_sample=0.0`), so I will just delete the check.

Reply:
I think there were other TF versions where this resulted in exceptions, or unexpected behavior, or also a tensor in graph mode (I definitely had this case).

Also, `a != 0.0` reads as if you are checking if it is not equal to 0.0 which is not true if this is a tensor.

If you want to check `not isinstance(use_sample, tf.Tensor) and use_sample != 0.0`, then write that (more explicitly).

Reply:
> I think there were other TF versions where this resulted in exceptions, or unexpected behavior, or also a tensor in graph mode

At least since 1.13 this works, but I understand the concerns about readability. I was just very confused because you mentioned I would do an `if` check directly on a tensor and I did not understand what you were getting at.